### PR TITLE
Optimize `Substs::super_fold_with`.

### PR DESCRIPTION
**Pull request by nnethercote**
This speeds up some of the rustc-benchmarks by up to ~4%.

```
futures-rs-test  4.467s vs  4.387s --> 1.018x faster (variance: 1.001x, 1.006x)
helloworld       0.242s vs  0.246s --> 0.980x faster (variance: 1.007x, 1.013x)
html5ever-2016-  7.664s vs  7.630s --> 1.004x faster (variance: 1.008x, 1.006x)
hyper.0.5.0      5.218s vs  5.133s --> 1.016x faster (variance: 1.013x, 1.008x)
inflate-0.1.0    5.040s vs  5.103s --> 0.988x faster (variance: 1.005x, 1.008x)
issue-32062-equ  0.361s vs  0.345s --> 1.047x faster (variance: 1.008x, 1.019x)
issue-32278-big  1.874s vs  1.850s --> 1.013x faster (variance: 1.020x, 1.018x)
jld-day15-parse  1.569s vs  1.508s --> 1.040x faster (variance: 1.009x, 1.003x)
piston-image-0. 12.210s vs 11.903s --> 1.026x faster (variance: 1.045x, 1.010x)
regex.0.1.30     2.568s vs  2.555s --> 1.005x faster (variance: 1.018x, 1.044x)
rust-encoding-0  2.139s vs  2.135s --> 1.001x faster (variance: 1.012x, 1.005x)
syntex-0.42.2   33.099s vs 32.353s --> 1.023x faster (variance: 1.003x, 1.028x)
syntex-0.42.2-i 17.989s vs 17.431s --> 1.032x faster (variance: 1.009x, 1.018x)
```

r? @eddyb. I don't know how this interacts with the changes that dikaiosune has been working on.

### CODE REVIEW

File: src/librustc/ty/subst.rs [hunk 1]
```diff
@@ -314,7 +314,37 @@ impl<'a, 'gcx, 'tcx> Substs<'tcx> {
...
+        //       params: params
+        //   })
+        //
+        // But this function is sufficiently hot that it's a sizeable win to
+        // handle some cases specially.
+        // - If folding doesn't change the substs, we avoid calling `mk_substs`
+        //   and instead reuse the existing substs.
+        // - In that case, when the substs length is one, we also avoid
+        //   creating a new Vec, which avoids a heap allocation.
+        let params;
```
Comment (KalitaAlexey) [inline src/librustc/ty/subst.rs]: `params` is used only in 338 line block. Can you move into the block?

File: src/librustc/ty/subst.rs [hunk 2]
```diff
@@ -314,7 +314,37 @@ impl<'a, 'gcx, 'tcx> Substs<'tcx> {
...
+        if self.params.len() == 1 {
```
Comment (KalitaAlexey) [inline src/librustc/ty/subst.rs]: Can you rewrite it to match self.params.len()?

File: src/librustc/ty/subst.rs (hunk 1 above)
Comment (nnethercote) [inline src/librustc/ty/subst.rs]: No. It's used on lines 336, 338, 349.

File: src/librustc/ty/subst.rs (hunk 2 above)
Comment (nnethercote) [inline src/librustc/ty/subst.rs]: I prefer having it this way. The `1` case is most frequent, then `>1`, and `0` is a distant last. With a match the ordering is less clear (i.e. not guaranteed).

File: src/librustc/ty/subst.rs [hunk 3]
```diff
@@ -314,7 +314,37 @@ impl<'a, 'gcx, 'tcx> Substs<'tcx> {
...
+            } else {
+                params = vec![param]
+            }
+        } else if self.params.len() > 1 {
+            params = self.params.iter().map(|k| k.fold_with(folder)).collect();
+            if params == self.params {
+                return &self;
+            }
+        } else {
+            return &self;
```
Comment (bluss) [inline src/librustc/ty/subst.rs]: super minor, but the `&` in `&self` is redundant. The `Self` type is `&'tcx Substs<'tcx>` and the `self` identifier is of type `&Self`. Deref coercion means that using `&self`, `self` or `*self` here is going to be equivalent.

**Comment by eddyb** [comment]
So this is an interesting take. I'd like to do it *after* moving to on-stack arrays and slice arena allocations instead of always allocating a `Vec`. Then it could either be an extension of that algorithm or maybe it wouldn't even have a measurable impact etc

**Comment by eddyb** [comment]
Oh and this doesn't handle `ty::relate` which may have a larger impact.

File: src/librustc/ty/subst.rs (hunk 2 above)
Comment (KalitaAlexey) [inline src/librustc/ty/subst.rs]: @nnethercote Okay. I got it.

**Comment by nnethercote** [comment]
> I'd like to do it after moving to on-stack arrays and slice arena allocations instead of always allocating a Vec.

My original version of the patch just checked if `params == self.params` and skipped the `mk_substs` call in that case, and that alone was a decent improvement because it avoids the interning, which is dominated by hash operations.

Then I added the skip-Vec-creation-in-the-length-1 case and that improved things a bit more.

**Comment by nnethercote** [comment]
> Oh and this doesn't handle ty::relate which may have a larger impact.

I haven't seen it in any of the profiles I've looked at. Can you describe in more detail how you think it might have an impact?

**Comment by nnethercote** [comment]
I have changed the code to just do the "skip mk_substs if the fold was a no-op" optimisation. That should avoid any conflicts with @Mark-Simulacrum's changes to `Substs`. It still gives some nice speed-ups.
```
futures-rs-test  4.189s vs  4.157s --> 1.008x faster (variance: 1.004x, 1.007x)
helloworld       0.223s vs  0.223s --> 1.000x faster (variance: 1.013x, 1.012x)
html5ever-2016-  5.192s vs  5.147s --> 1.009x faster (variance: 1.007x, 1.004x)
hyper.0.5.0      4.788s vs  4.763s --> 1.005x faster (variance: 1.008x, 1.008x)
inflate-0.1.0    4.553s vs  4.494s --> 1.013x faster (variance: 1.010x, 1.044x)
issue-32062-equ  0.324s vs  0.314s --> 1.032x faster (variance: 1.008x, 1.020x)
issue-32278-big  1.661s vs  1.654s --> 1.005x faster (variance: 1.008x, 1.002x)
jld-day15-parse  1.475s vs  1.422s --> 1.038x faster (variance: 1.007x, 1.008x)
piston-image-0. 11.386s vs 11.227s --> 1.014x faster (variance: 1.005x, 1.063x)
reddit-stress    2.359s vs  2.308s --> 1.022x faster (variance: 1.017x, 1.024x)
regex.0.1.30     2.421s vs  2.388s --> 1.014x faster (variance: 1.036x, 1.008x)
rust-encoding-0  1.983s vs  1.973s --> 1.005x faster (variance: 1.008x, 1.016x)
syntex-0.42.2   29.582s vs 29.610s --> 0.999x faster (variance: 1.014x, 1.045x)
syntex-0.42.2-i 15.203s vs 15.175s --> 1.002x faster (variance: 1.011x, 1.006x)
```
r? @eddyb

**Comment by Mark-Simulacrum** [comment]
Note that it will be at least a few days, more likely a week until I can resume work on the `Substs` improvements, so I'm not sure removing things now is a good idea. I haven't followed this PR closely though, not sure how large/significant the changes here are.

File: src/librustc/ty/subst.rs [hunk 4]
```diff
@@ -304,8 +304,15 @@ impl<'a, 'gcx, 'tcx> Substs<'tcx> {
 
 impl<'tcx> TypeFoldable<'tcx> for &'tcx Substs<'tcx> {
     fn super_fold_with<'gcx: 'tcx, F: TypeFolder<'gcx, 'tcx>>(&self, folder: &mut F) -> Self {
-        let params = self.iter().map(|k| k.fold_with(folder)).collect();
-        folder.tcx().mk_substs(params)
+        let params: Vec<_> = self.iter().map(|k| k.fold_with(folder)).collect();
+
+        // If folding doesn't change the substs, it's faster to avoid calling
+        // `mk_substs` and instead reuse the existing substs.
```
Comment (eddyb) [inline src/librustc/ty/subst.rs]: Hmm, maybe open an issue about doing this in general, or at least further investigation? That is, there's a bunch of places where we can short-circuit the interner, maybe some are worth doing.

**Review by eddyb — APPROVED** [review]

**Comment by alexcrichton** [comment]
@bors: retry force clean

* restarted buildbot

**Pull request merged**